File: scripts/tests/test_raceresult_parse.js
/**
 * worker.js parseRaceresultHTML の回帰テスト (2026-07-19 markup 変更対応)。
 *   boatrace.jp が払戻券種ラベルを th→td rowspan に変更し旧パーサが全滅した
 *   実障害の固定。フィクスチャは probe の実採取 HTML (場2 戸田 1R 20260719)。
 *
 *   node scripts/tests/test_raceresult_parse.js
 */
const assert = require('assert');
const fs = require('fs');
const path = require('path');
const vm = require('vm');

const src = fs.readFileSync(path.join(__dirname, '..', '..', 'cloudflare-worker', 'worker.js'), 'utf8');

// worker.js から必要な関数だけ brace-matching で抽出して eval する
function extractFn(name) {
  const i = src.indexOf(`function ${name}(`);
  assert.ok(i >= 0, `function ${name} not found in worker.js`);
  let depth = 0;
  const j = src.indexOf('{', i);
  for (let k = j; k < src.length; k++) {
    if (src[k] === '{') depth++;
    else if (src[k] === '}') { depth--; if (depth === 0) return src.slice(i, k + 1); }
  }
  throw new Error(`unbalanced braces for ${name}`);
}

const ctx = { console };
vm.createContext(ctx);
for (const fn of ['stripTags', 'toHalfWidthInt', 'parseRaceresultHTML']) {
  vm.runInContext(extractFn(fn), ctx);
}

const html = fs.readFileSync(path.join(__dirname, 'fixtures', 'raceresult_new_markup.html'), 'utf8');
// vm 別レルムの Object と deepStrictEqual の prototype 比較を避けるため JSON 正規化
const r = JSON.parse(JSON.stringify(vm.runInContext('parseRaceresultHTML(HTML, 2, 1, "2026-07-19")',
  Object.assign(ctx, { HTML: html }))));

let pass = 0, fail = 0;
function t(name, f) {
  try { f(); pass++; console.log(`  PASS: ${name}`); }
  catch (e) { fail++; console.log(`  FAIL: ${name} — ${e.message}`); }
}

console.log('[parseRaceresultHTML: 2026-07-19 markup]');
t('着順 6 艇 (全角数字対応)', () => {
  assert.strictEqual(r.boats.length, 6);
  const byPlace = {};
  r.boats.forEach(b => { byPlace[b.racer_place_number] = b.racer_boat_number; });
  assert.deepStrictEqual(byPlace, { 1: 4, 2: 2, 3: 3, 4: 1, 5: 5, 6: 6 });
});
t('確定判定 (technique=1)', () => assert.strictEqual(r.race_technique_number, 1));
t('3連単 (td rowspan ラベル + span 組番 + &yen; 金額)', () =>
  assert.deepStrictEqual(r.payouts.trifecta, [{ combination: '4-2-3', amount: 5350 }]));
t('2連単 / 単勝', () => {
  assert.deepStrictEqual(r.payouts.exacta, [{ combination: '4-2', amount: 1930 }]);
  assert.deepStrictEqual(r.payouts.win, [{ combination: '4', amount: 860 }]);
});
t('3連複 同着 (rowspan 継続行)', () =>
  assert.deepStrictEqual(r.payouts.trio, [
    { combination: '2=3=4', amount: 1020 },
    { combination: '1=2=4', amount: 980 },
  ]));
t('&nbsp; 埋め草行の除外', () => {
  assert.strictEqual(r.payouts.trifecta.length, 1);
  assert.strictEqual(r.payouts.win.length, 1);
});
t('未掲載ページは technique=null', () => {
  const empty = vm.runInContext(
    'parseRaceresultHTML("<table><tbody><tr><td>データはありません</td></tr></tbody></table>", 2, 1, "2026-07-19")', ctx);
  assert.strictEqual(empty.race_technique_number, null);
});

console.log(`\n${pass} passed, ${fail} failed`);
process.exit(fail > 0 ? 1 : 0);

File: scripts/tests/test_raceresult_parse.py
"""raceresult パーサの回帰テスト (2026-07-19 markup 変更対応の固定)。

実障害: boatrace.jp が raceresult の markup を変更 (テーブル class 刷新 +
払戻券種ラベル th→td rowspan) し、旧 .table1 / th 前提のパースが全滅。
07-18 は archive 180 レース全て finished=0、Worker は openapi ミラー欠落場
(07-19 の場5) で「着順のみ・払戻なし」が数十分継続。
フィクスチャは probe の実採取 HTML (場2 戸田 1R 20260719) を縮小したもの。
"""
import os
import sys
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

try:
    import scrape_results as SR
    from bs4 import BeautifulSoup  # noqa: F401
    _HAVE_DEPS = True
except Exception:
    _HAVE_DEPS = False

_FIXTURE = os.path.join(os.path.dirname(__file__), "fixtures", "raceresult_new_markup.html")


@unittest.skipUnless(_HAVE_DEPS, "bs4 missing")
class TestRaceresultParse(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        with open(_FIXTURE, encoding="utf-8") as f:
            cls.html = f.read()
        cls.r = SR.parse_raceresult(cls.html, 2, 1)

    def test_boats_all_six(self):
        self.assertEqual(len(self.r["boats"]), 6)
        by_place = {b["racer_place_number"]: b["racer_boat_number"] for b in self.r["boats"]}
        self.assertEqual(by_place, {1: 4, 2: 2, 3: 3, 4: 1, 5: 5, 6: 6})

    def test_finished(self):
        self.assertEqual(self.r["race_technique_number"], 1)

    def test_racer_name_stripped(self):
        first = [b for b in self.r["boats"] if b["racer_place_number"] == 1][0]
        # 登番 4444 は除去され氏名だけ残る
        self.assertNotIn("4444", first["racer_name"])
        self.assertIn("ヤマダ", first["racer_name"])

    def test_trifecta_new_markup(self):
        # 券種ラベルが td rowspan (旧 th ではない) でも抽出できる
        self.assertEqual(self.r["payouts"]["trifecta"],
                         [{"combination": "4-2-3", "amount": 5350}])

    def test_exacta_and_win(self):
        self.assertEqual(self.r["payouts"]["exacta"],
                         [{"combination": "4-2", "amount": 1930}])
        self.assertEqual(self.r["payouts"]["win"],
                         [{"combination": "4", "amount": 860}])

    def test_trio_dead_heat_continuation(self):
        # rowspan 継続行 (同着 2 本目、ラベル無し) が直前券種に入る
        self.assertEqual(self.r["payouts"]["trio"],
                         [{"combination": "2=3=4", "amount": 1020},
                          {"combination": "1=2=4", "amount": 980}])

    def test_nbsp_filler_rows_excluded(self):
        # &nbsp; 埋め草行が混入しない (各券種の件数が実データ数と一致)
        self.assertEqual(len(self.r["payouts"]["trifecta"]), 1)
        self.assertEqual(len(self.r["payouts"]["win"]), 1)

    def test_unfinished_page_returns_none(self):
        # 結果未掲載 (着順 tbody 無し) は technique=None で汚染しない
        r = SR.parse_raceresult("<html><body><table><tbody><tr><td>データはありません</td></tr></tbody></table></body></html>", 2, 1)
        self.assertIsNone(r["race_technique_number"])
        self.assertEqual(r["boats"], [])


if __name__ == "__main__":
    unittest.main()
